### ISSUE TITLE
[docs] Improve documentation for LayerNorm

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#51455]
https://pytorch.org/docs/stable/generated/torch.nn.LayerNorm.html: `The mean and standard-deviation are calculated separately over the last certain number dimensions which have to be of the shape specified by normalized_shape`

What's not clear from this excerpt (and the rest of the text):

1. What dimensions is the mean computed over? What is the shape of the computed mean/std? If it's the last few dimensions as suggested by this text, then does it mean that for BCHW layout, height and width must be fixed and passed at LayerNorm construction time? This seems wrong. Also expression "calculated separately over the last certain number of dims" may imply exactly the reverse: that the first dims are aggregated over. What seems to be happening is that "mean and standard-deviation are calculated over all except the first dimension (assumed batch dimension) and the specified last dimension (assumed channels dimensions)".
2. If the last few dimensions (including spatial/temporal) are aggregated over, why does LayerNorm need their (potentially dynamic) shape at construction time?

2. Is there any layout are assumed for a typical usecase (with fixed number of channels and various time/spatial dimension)? BTC? BCT? This is important, because PyTorch usually uses BC... logical layouts in other contexts

3. Could it include a visual explanation of what dimensions are aggregated over?

The examples given in the docs don't include dimension semantics (such as Channels, Time, Height, Width), so I couldn't extract any information from them.

E.g. an example of type:
```python
import torch

ln = torch.nn.LayerNorm(256, elementwise_affine = False)

# BTC
a = torch.randn(8, 1000, 256) * 100 + 200
print(a[0, ...].mean(), a[0, ...].std())
#tensor(199.9173) tensor(99.8517)

print(a[0, 0].mean(), a[0, 0].std())
#tensor(206.8604) tensor(104.7011)

b = ln(a)
print(b[0, ...].mean(), b[0, ...].std())
# tensor(2.2389e-09) tensor(1.0000)
print(a[0, 0].mean(), a[0, 0].std())
#tensor(-2.6077e-08) tensor(1.0020)

# what's going on?
# Does LayerNorm standardize every (b, ...) slice as illustrated in GroupNorm paper's figure?
# Does LayerNorm standardizes every (b, t, :) slice (i.e. standardizes embedding vectos themselves - does it even make sense)?
```
is much clearer.

Phrasing "over the last certain number dimensions" seems very ambiguous, as the aggregation seems to happen actually over not specified dimensions.

cc @jlin27 @mruberry @albanD @jbschlosser

**Comment by vadimkantorov** [comment]
Theoretically, this `normalized_shape` is not used by the layer norm itself, e.g. as seen in https://github.com/pytorch/pytorch/blob/4a0d17ba2d5966cea27dbce527baef62bb50c6f7/torch/csrc/jit/passes/decompose_ops.cpp#L209, only the number of last dimensions matters

**Comment by vadimkantorov** [comment]
This is also importing for porting TF code in PyTorch. E.g. LayerNormalization https://www.tensorflow.org/api_docs/python/tf/keras/layers/LayerNormalization only requires to specify the dimensions, not their shapes.

Also, it says: "Given a tensor inputs, moments are calculated and normalization is performed across the axes specified in axis.". (axis by default is set to -1). Is TF's instance of `LayerNormalization()` with default parameters expressable as PyTorch LayerNorm?

In TF it's also not clear whether it does flattens/aggregates only specified dimensions, all except them, or all dimensions except batch dim.

**Comment by vadimkantorov** [comment]
Why does LayerNorm need the full shape to be normalized over? This seems to block variable-sized batches?

And for the BTC case, does `LayerNorm(C)` in fact  do `(x - x.mean(dim = -1)) / (x.var(dim = -1) + eps).sqrt()`? in the sense that separate mean/var statistics is used per b, h, w. This does not seem to even match any of traditional variants including InstanceNorm (which would compute separate stats per b, c) :)
![image](https://user-images.githubusercontent.com/1041752/106491728-73d40680-64b7-11eb-8769-3f758996e959.png)

**Comment by vadimkantorov** [comment]
It seems that PyTorch's nn.LayerNorm is doing:
![image](https://user-images.githubusercontent.com/1041752/106503787-62dec180-64c6-11eb-9baf-dd91c86061dd.png)

**Comment by vadimkantorov** [comment]
@ezyang Could you please reopen this issue?

The standing items:
- add pseudocode
- add a realistic vision example or remove the BCHW example (with fixed image size)

**Comment by vadimkantorov** [comment]
Maybe, also NaiveBatchNorm / SyncBatchNorm pseudo-code or simple-code could be presented in the docs, as implemented by https://detectron2.readthedocs.io/en/latest/_modules/detectron2/layers/batch_norm.html#NaiveSyncBatchNorm

**Comment by yuzhms** [comment]
Hi, I am using the layer normalization in image. I am curious about the exact behavior that the `nn.LayerNorm` did. If I pass `normalized_shape=channel_dim` to nn.LayerNorm, does it perform the Layernorm as described in GroupNorm's paper as:
![image](https://user-images.githubusercontent.com/30227789/137616690-bff425a8-fddb-4991-9976-9f529088f40a.png)

or only calculating the mean and variance on the single channel dimension as you mentioned in 

> It seems that PyTorch's nn.LayerNorm is doing: ![image](https://user-images.githubusercontent.com/1041752/106503787-62dec180-64c6-11eb-9baf-dd91c86061dd.png)

Thanks very much if someone could make it clear. 

Also, since Pytorch hides the implementation of LayerNorm out of Python. It will be nice that a pseudo-code illustration could be added in the documentation.
@vadimkantorov

**Comment by vadimkantorov** [comment]
In my understanding, LayerNorm in groupnorm paper has little to do how it's defined in PyTorch :( And it assumes B..C shape (as opposed to BCHW)

**Comment by vadimkantorov** [comment]
It seems that decomps are now placed into torch._decomp, these could actually be identical to "reference implementations" and could be good snippets for explaining their function for documentation

**Comment by mruberry** [comment]
> It seems that decomps are now placed into torch._decomp, these could actually be identical to "reference implementations" and could be good snippets for explaining their function for documentation

(see torch/_refs/, where we're working on moving them)

**Comment by digital-idiot** [comment]
> Also, it says: "Given a tensor inputs, moments are calculated and normalization is performed across the axes specified in axis.". (axis by default is set to -1). Is TF's instance of LayerNormalization() with default parameters expressable as PyTorch LayerNorm?

I am also interested in this. In some cases explicitly specifying exact shape can be difficult. I prefer the `tensorflow/keras`'s approach of specifying the `axis` instead.
